nbdime/gitfiles.py

```diff
@@ -93,15 +100,15 @@ def is_path_in_repo(path):
         return False
 
 
-def _get_diff_entry_stream(path, blob, ref_name, repo_dir, special_ref_name):
+def _get_diff_entry_stream(path, blob, ref_name, repo_dir):
     """Get a stream to the notebook, for a given diff entry's path and blob
 
     Returns None if path is not a Notebook file, and EXPLICIT_MISSING_FILE
     if path is missing."""
     if path:
         if not path.endswith('.ipynb'):
             return None
-        if special_ref_name == 'WORKING':
+        if blob is None:
```

Comment:
There's an issue here while performing the diff between two Git refs,  where a file was added. The Base always points to the "Working Tree" version, whereas it should be empty,  so when the client computes the diff on top of the base it shows an incorrect result. I attempted to fix this in the commit https://github.com/jupyter/nbdime/pull/482/commits/ef13feb1bf511a724e437bf35b3c3bc782b0cf06. I'll try and have a stab at fixing this over your commits.

Reply:
![](https://user-images.githubusercontent.com/43826141/62234924-96e0a000-b380-11e9-9c89-492c505c2848.png)

Reply:
I don't really understand this issue. In the docs you link in the commit (https://gitpython.readthedocs.io/en/stable/reference.html#git.diff.Diff), it says that for added/deleted files the `path` will be `None`. We check if the `path` is `None` at the start of the function. According to the docs, if the path is not `None` and the `blob` is, then we are diffing against the working tree.

Would you mind sharing a minimal reproducible sample? I don't really understand the screenshot..

Reply:
@vidartf 

Repro setup:

* Created this repo for the repo case https://github.com/jaipreet-s/diff-harness/commits/master
* I have some changes in my Working Tree, `print('Simple Diff Committed` => `print('Simple Diff in Working Tree)`
* We're trying to diff between commits eaf1741 and b82e191 in the repo so the API input is [1].

Repro output:
*  In the current state of the API (at commit https://github.com/jupyter/nbdime/commit/7bb092833ae3271c5d8fbd66e9d82cccf9362abd) the output is as expected with base file empty  [here](https://gist.github.com/jaipreet-s/396da4a2c13f476dec393e480b4e8b20#file-correct ).
*  In the state at https://github.com/jupyter/nbdime/pull/482/commits/94100ab615efad1fd388bc230306f3dd0b290cf7, the "base" file is the state in the Working Tree [here](https://gist.github.com/jaipreet-s/396da4a2c13f476dec393e480b4e8b20#file-incorrect)

So the issue is about the "base" output being returned incorrectly if we're trying to diff a commit in the history where a file has been added/

[1] 
```
{
	"ref_local": {
		"git": "eaf1741"
		
	},
	"ref_remote": {
		"git": "b82e191"
	},
	"file_path": "SimpleDiff.ipynb"
}
```

Reply:
>  According to the docs, if the path is not None and the blob is, then we are diffing against the working tree.

I've documented the values for path, blob, and ref_name for 3 cases https://gist.github.com/jaipreet-s/396da4a2c13f476dec393e480b4e8b20#file-ref-blob-path-values

* _eaf1741 vs b82e191_: blob is `None` in `eaf1741` since the file was first added in `b82e191`. ref_name is the Git ref in both cases.
* _HEAD vs WORKING_: blob is `6ee0` for HEAD. For Working, ref_name **as well as** blob is None.
* _HEAD vs INDEX:_ blob is 6ee0 for HEAD. For Index, ref_name is `Diffable.Index` and blob is `58d9d9`

So, the issue is coming from the fact that `blob` can be `None` both in the Working Tree and if file is added/removed in a Git ref. But only the Working Tree will have the `ref_name` as None. 

That is why in https://github.com/jupyter/nbdime/commit/7bb092833ae3271c5d8fbd66e9d82cccf9362abd I've checked if `ref_name` is the `WorkingTree` explicitly and only then use the content from the Working Tree.

Reply:
Ok, I see now on the GitPython repo there is an issue about this https://github.com/gitpython-developers/GitPython/issues/749 .